### PR TITLE
Emulator backend API support

### PR DESCRIPTION
**Pull request by keith-zephyr**
Add a mechanism for emulators of the same device type to declare a common backend API. 

See [RFC #53791](https://github.com/zephyrproject-rtos/zephyr/issues/53791) for more information.

### CODE REVIEW

File: include/zephyr/drivers/emul_stub_device.h [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+/* For every instance of a DT_DRV_COMPAT stub out a device for that instance */
+#define EMUL_STUB_DEVICE(n)                                                                        \
+	__maybe_unused static int emul_init_stub_##n(const struct device *dev)                     \
+	{                                                                                          \
+		ARG_UNUSED(dev);                                                                   \
+		return 0;                                                                          \
+	}                                                                                          \
+                                                                                                   \
+	/* Since this is only stub, allocate the structs once. */                                  \
```
Comment (alevkoy) [inline include/zephyr/drivers/emul_stub_device.h:35]: Would they be allocated more times if this wasn't a stub?

File: include/zephyr/drivers/emul.h [hunk 2]
```diff
@@ -120,8 +122,9 @@ extern const struct emul __emul_list_end[];
  * @param data_ptr emulator-specific data
  * @param cfg_ptr emulator-specific configuration data
  * @param bus_api emulator-specific bus api
+ * @param backend_api emulator-specific backend api
  */
-#define EMUL_DT_DEFINE(node_id, init_fn, data_ptr, cfg_ptr, bus_api)                               \
+#define EMUL_DT_DEFINE(node_id, init_fn, data_ptr, cfg_ptr, bus_api, _backend_api)                 \
```
Comment (alevkoy) [inline include/zephyr/drivers/emul.h:127]: Why the leading underscore?

File: tests/subsys/emul/src/emul_test_driver.c [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+static int emul_test_driver_init(const struct emul *target, const struct device *parent)
+{
+	ARG_UNUSED(target);
+	ARG_UNUSED(parent);
+
+	return 0;
+}
+
+#define EMUL_TEST_DRIVER(n)                                                                        \
```
Comment (alevkoy) [inline tests/subsys/emul/src/emul_test_driver.c:72]: I can't tell whether this is a structural issue or just semantics, but AFAICT, the point of the stub driver is to satisfy the requirement for a linked `struct device` for `struct emul`s that don't have a natural one. If that's true, then it seems to me like the names of this module and its members should just be "emul_something" and not "something_driver." The test is written in terms of the backend API, which is an emulator API. From the point of view of other modules, this module provides an emulator, and the fact that it contains a driver is an implementation detail.

**Review by aaronemassey — CHANGES_REQUESTED** [review]
I quite like this for making it easier to write API specific tests instead of chip/driver specific tests. The API can then easily be tested with various drivers/chips but the test can remain the same.

**Review by aaronemassey — DISMISSED** [review]

File: include/zephyr/drivers/emul_stub_device.h (hunk 1 above)
Comment (alevkoy) [reply include/zephyr/drivers/emul_stub_device.h:35]: Are you going to get rid of it?

**Review by alevkoy — COMMENTED** [review]
Looks good except for the comment that still doesn't make sense.

**Comment by keith-zephyr** [comment]
Note - commit https://github.com/zephyrproject-rtos/zephyr/commit/774949c7346a0e3984beb8aa485741f08fcb3f96 is a rebase only.

**Comment by keith-zephyr** [comment]
> Looks good except for the comment that still doesn't make sense.

Fixed.

**Review by alevkoy — DISMISSED** [review]

**Comment by keith-zephyr** [comment]
@yperess Can you give this a review?

**Review by yperess — CHANGES_REQUESTED** [review]
Please add some details of this to https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/hardware/emulator/index.rst. I think it'll be interesting to show an example like we discussed offline where we want to run the same test using 2 different drivers of a similar "class" and allowing the same test to exercise both drivers.

**Review by aaronemassey — DISMISSED** [review]

**Review by yperess — APPROVED** [review]

**Review by fabiobaltieri — APPROVED** [review]

**Pull request merged**